### PR TITLE
W-11706176: Enrich ApplicationTypeLoader with the types from the dependencies

### PR DESCRIPTION
**Pull request by eze210**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java [hunk 1]
```diff
@@ -23,15 +38,61 @@ public class ApplicationTypeLoader implements TypeLoader {
...
+    if (special.isPresent()) {
+      return special;
+    }
+
+    return getFromDependency(typeIdentifier);
+  }
+
+  // The string format can be the full name of the type, or a string with syntax <extension-prefix>:<type-alias> if
+  // the type has an alias. This format is temporal in order to use the types from the test operations, but may change
+  // when the type catalog using dataweave is fully implemented.
```
Comment (mbuchwald) [inline core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java:70]: Do we have a ticket to reference to that?

File: core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java [hunk 2]
```diff
@@ -23,15 +38,61 @@ public class ApplicationTypeLoader implements TypeLoader {
...
+  }
+
+  private static boolean matchesAlias(String expectedAlias, MetadataType metadataType) {
+    Optional<TypeAliasAnnotation> aliasAnnotation = metadataType.getAnnotation(TypeAliasAnnotation.class);
+    if (aliasAnnotation.isPresent()) {
+      return aliasAnnotation.get().getValue().equals(expectedAlias);
     } else {
-      return specialTypesLoader.load(typeIdentifier);
+      return false;
     }
```
Comment (mbuchwald) [inline core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java:96]: I would change this to: 
```java
return aliasAnnotation.map(annotation -> annotation.getValue().equals(expectedAlias)).orElse(false);
```

File: core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java (hunk 1 above)
Comment (eze210) [reply core/src/main/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoader.java:70]: Added the reference

File: core-tests/src/test/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoaderTestCase.java [hunk 3]
```diff
@@ -24,11 +40,29 @@
...
+
+    XmlDslModel dslModel = XmlDslModel.builder().setPrefix(MOCK_EXTENSION_PREFIX).build();
+    when(mockExtensionModel.getXmlDslModel()).thenReturn(dslModel);
+
+    when(mockExtensionModel.getName()).thenReturn(MOCK_EXTENSION_NAME);
+
+    Map<Class<? extends TypeAnnotation>, TypeAnnotation> annotationMap = new HashMap<>(2);
+    annotationMap.put(TypeAliasAnnotation.class, new TypeAliasAnnotation(MOCK_TYPE_ALIAS));
+    annotationMap.put(TypeIdAnnotation.class, new TypeIdAnnotation(FULL_CLASS_NAME_FOR_MOCK_TYPE));
+    ObjectType mockType = new DefaultObjectType(emptyList(), false, null, JAVA, annotationMap);
```
Comment (asanguinetti) [inline core-tests/src/test/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoaderTestCase.java:62]: I believe there are builders for these things. Shouldn't we use them instead?

Should look like this I believe:
```suggestion
    BaseTypeBuilder.create(JAVA).objectType()
      .id(FULL_CLASS_NAME_FOR_MOCK_TYPE)
      .with(new TypeAliasAnnotation(MOCK_TYPE_ALIAS))
      .build()
```

**Review by asanguinetti — COMMENTED** [review]
Just a small suggestion.
Otherwise LGTM.

File: core-tests/src/test/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoaderTestCase.java (hunk 3 above)
Comment (eze210) [reply core-tests/src/test/java/org/mule/runtime/core/api/type/catalog/ApplicationTypeLoaderTestCase.java]: Done

**Review by asanguinetti — APPROVED** [review]

**Comment by eze210** [comment]
--validate

**Review by asanguinetti — APPROVED** [review]

**Comment by eze210** [comment]
--validate

**Comment by eze210** [comment]
--validate

**Review by mbuchwald — APPROVED** [review]

**Comment by eze210** [comment]
--validate

**Pull request merged**